sherpa/tests/test_sherpa_io.py

```diff
@@ -0,0 +1,304 @@
+    (colnames, coldata, fname) = get_ascii_data(str(tfile), ncols=1)
+    assert colnames == ["col2", "xcol", "ycol"]
+    assert len(coldata) == 1
+    assert coldata[0] == pytest.approx([-0.2, -0.4])
+
+
+def test_get_ascii_data_more_colnames_than_columns_no_colkeys_ncols2(tmp_path):
+    """What happens if more column names than columns: ncols=2?"""
+
+    tfile = tmp_path / "col.dat"
+    tfile.write_text("# col2 xcol ycol\n-0.2 23\n-0.4 24\n")
+
+    (colnames, coldata, fname) = get_ascii_data(str(tfile), ncols=2)
+    assert colnames == ["col2", "xcol", "ycol"]
+    assert len(coldata) == 2
+    assert coldata[0] == pytest.approx([-0.2, -0.4])
+    assert coldata[1] == pytest.approx([23, 24])
+
+
+def test_get_ascii_data_more_colnames_than_columns_colkeys_ncols1(tmp_path):
+    """What happens if more column names than columns? colkeys/ncols=1
+
+    There is also a check for specifying an invalid column name but
+    that check is not reached.
+    """
+
+    tfile = tmp_path / "col.dat"
+    tfile.write_text("# col2 xcol ycol\n-0.2 23\n-0.4 24\n")
+
+    # Although ncols=1 is the default be explicit in this test
+    with pytest.raises(IOErr,
+                       match="^Expected 2 columns but found 3$"):
+        get_ascii_data(str(tfile), ncols=1, colkeys=["col2", "col3"])
+
+
+def test_get_ascii_data_more_colnames_than_columns_colkeys_ncols2(tmp_path):
+    """What happens if more column names than columns? colkebys/ncols=2
+
+    There is also a check for specifying an invalid column name but
+    that check is not reached.
+    """
+
+    tfile = tmp_path / "col.dat"
+    tfile.write_text("# col2 xcol ycol\n-0.2 23\n-0.4 24\n")
+
+    with pytest.raises(IOErr,
+                       match="^Expected 2 columns but found 3$"):
+        get_ascii_data(str(tfile), ncols=2, colkeys=["col2", "col3"])
+
+
+def test_get_ascii_data_too_many_colkeys(tmp_path):
+    """Regression test"""
+
+    tfile = tmp_path / "1col.dat"
+    tfile.write_text("23 400\n24 -0.2\n")
+
+    with pytest.raises(IOErr,
+                       match=r"Required column 'col3' not found in \['col1'. 'col2'\]"):
+        get_ascii_data(str(tfile), ncols=3, colkeys=["col1", "col2", "col3"])
+
+
+def test_get_ascii_data_unknown_colkey(tmp_path):
+    """Regression test"""
+
+    tfile = tmp_path / "1col.dat"
+    tfile.write_text("# col2\n23\n24\n")
+
+    with pytest.raises(IOErr,
+                       match=r"Required column 'col1' not found in \['col2'\]"):
+        get_ascii_data(str(tfile), colkeys=["col1", "col2", "col3"])
+
+
+def test_write_arrays_single_column(tmp_path):
+    """What happens if do not give a list of lists?
+
+    Should this error out? Users may be relying on this behavior, so
+    we do not error out, but that seems unlikely.
+
+    """
+
+    tfile = tmp_path / "not-changed.dat"
+    with pytest.raises(IOErr,
+                       match=r"^please supply array\(s\) to write to file$"):
```

Comment:
I like how the error message politely says "please" and not just "XYZ IS REQUIRED, YOU FOOL"

Reply:
We can put that in a later PR....